### PR TITLE
Remove the lexer hack in Painless

### PR DESCRIPTION
**Pull request by jdconrad**
Background: 

The lexer-hack (https://en.wikipedia.org/wiki/Lexer_hack) was used in Painless to add contextual information around types during lexing. This made the the lexer/grammar much simpler, and allowed us to build a user tree with full typing information at creation.

Limitations: 

Forcing the lexer to know about type information up front requires  that we know all types when building the user tree, so this limits us from a number of things including the following:

1. good, consistent error messages as types fail during lexing which gives very poor error messaging since this should be purely syntax errors
2. compiling against different contexts as a possible use case for ensuring a stored script is valid for at least one context
3. pursuing new front ends such as a possible template language based on Painless
4. using the existing grammar to introduce an auto-completion API
5. the potential of adding new features with specialized types.

Results:

This change removes the lexer hack by doing all type checking as part of the semantic pass against the user tree. The lexer now keywords on the primitives and def types while the user tree will use a symbol node instead of a variable node that can potentially be either a type, a piece of a package name, or a variable. One important change is that the dot operator is now overridden. This means that a dot operator may indicate a piece of a package name or a field (`a.b` could be access field `b` from variable `a` or could be the static type `a.b`). Logic has been introduced to account for this overloaded operator including checks in the expression nodes to ensure they aren't mistakenly using an unexpected static type or an unknown partial type.

Note this PR appears to be huge, but about 3/4 of it is changes to the auto-generated antlr lexer/parser and additional test cases.

### CODE REVIEW

File: modules/lang-painless/src/main/antlr/PainlessLexer.g4 [hunk 1]
```diff
@@ -120,23 +109,19 @@ INTEGER: ( '0' | [1-9] [0-9]* ) [lLfFdD]?;
...
+PRIMITIVE: 'boolean' | 'byte' | 'short' | 'char' | 'int' | 'long' | 'float' | 'double';
+DEF: 'def';
+
 ID: [_a-zA-Z] [_a-zA-Z0-9]*;
 
 mode AFTER_DOT;
 
-DOTINTEGER: ( '0' | [1-9] [0-9]* )                        -> mode(DEFAULT_MODE);
-DOTID: [_a-zA-Z] [_a-zA-Z0-9]*                            -> mode(DEFAULT_MODE);
+DOTINTEGER: ( '0' | [1-9] [0-9]* )                                            -> mode(DEFAULT_MODE);
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/antlr/PainlessLexer.g4:126]: Why is this indented                                                             so far?

File: modules/lang-painless/src/main/antlr/PainlessParser.g4 [hunk 2]
```diff
@@ -87,32 +87,38 @@ declaration
...
     : ID ( ASSIGN expression )?
     ;
 
 trap
-    : CATCH LP TYPE ID RP block
+    : CATCH LP type ID RP block
     ;
 
 noncondexpression
-    :               unary                                                               # single
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/antlr/PainlessParser.g4:102]: spaces?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Operation.java [hunk 3]
```diff
@@ -51,9 +51,7 @@
     EQ      ( "=="  , "equals"                 ),
     EQR     ( "===" , "reference equals"       ),
     NE      ( "!="  , "not equals"             ),
-    NER     ( "!==" , "reference not equals"   ),
-    INCR    ( "++"  , "increment"              ),
-    DECR    ( "--"  , "decrement"              );
+    NER     ( "!==" , "reference not equals"   );
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Operation.java:54]: Why is `INCR` and `DECR` removed here?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java [hunk 4]
```diff
@@ -627,12 +629,25 @@ public ANode visitBool(BoolContext ctx) {
...
         } else {
-            throw location(ctx).createError(new IllegalStateException("Illegal tree structure."));
+            throw location(ctx).createError(new IllegalStateException("illegal tree structure"));
         }
 
         return new EBool(location(ctx), operation, left, right);
     }
 
+    @Override
+    public ANode visitElvis(ElvisContext ctx) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java:639]: Why was this moved?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java [hunk 5]
```diff
@@ -43,27 +45,52 @@ public EVariable(Location location, String name) {
...
+
+            StaticNode staticNode = new StaticNode();
 
-        output.actual = variable.getType();
+            staticNode.setLocation(location);
+            staticNode.setExpressionType(output.actual);
 
-        VariableNode variableNode = new VariableNode();
+            output.expressionNode = staticNode;
+        } else if (scope.isVariableDefined(name)) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java:71]: Say you have a variable with same as package name, eg `java`, what happens?

Comment (stu-elastic) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java:71]: eg `def java = 1; java.lang.Integer`

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java [hunk 6]
```diff
@@ -43,27 +45,52 @@ public EVariable(Location location, String name) {
 
     @Override
     Output analyze(ClassNode classNode, ScriptRoot scriptRoot, Scope scope, Input input) {
-        if (input.read == false && input.write == false) {
-            throw createError(new IllegalArgumentException("not a statement: variable [" + name + "] not used"));
-        }
-
         Output output = new Output();
+        Class<?> type = scriptRoot.getPainlessLookup().canonicalTypeNameToType(name);
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java:49]: Is there any possibility of a partial name triggering `type != null`? eg. `java.util.HashMap.abc123`.

File: modules/lang-painless/src/main/antlr/PainlessLexer.g4 (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/antlr/PainlessLexer.g4:126]: Fixed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Operation.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Operation.java:54]: These are replaced with ADD/SUB and the Walker adds a constant of 1 in the appropriate places. The Painless user tree does not need to know about these natively since they are a shortcut. It only needs to know if a value from a variable being assigned to must be read from before or after the assignment.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java:639]: Rules got moved around here to stay in line with the ordering in the grammar.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java (hunk 6 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java:49]: Tests should cover all use cases of this particular corner case. This should always result in a compiler error - "not a symbol"

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java:71]: This `int java; java.util.Integer.parseInt('1')` results in an error since the java package name is overridden by a variable name.

This `int java; java.util.ArrayList x = new java.util.ArrayList()` is successful as our parser can recognize that declarations must be a type, so it uses a different parsing path based on the rules.

This `int java; List x = new java.util.ArrayList(); ArrayList y; y = (java.util.ArrayList)x;` is successful for the same reasons as the previous example.

These all follow similar behavior to Java with error messages differing based on our use cases.

**Review by stu-elastic — APPROVED** [review]
Really great work, Jack.

**Comment by jdconrad** [comment]
@stu-elastic Thank you for the review!
